pilot/pkg/model/network.go

```diff
@@ -108,40 +159,103 @@ func NewNetworkManager(env *Environment) *NetworkManager {
 		lcmVal = lcm(lcmVal, num)
 	}
 
-	return &NetworkManager{
-		lcm:                 uint32(lcmVal),
-		byNetwork:           byNetwork,
-		byNetworkAndCluster: byNetworkAndCluster,
+	mgr.lcm = uint32(lcmVal)
+	mgr.byNetwork = byNetwork
+	mgr.byNetworkAndCluster = byNetworkAndCluster
+
+	return gatewaySet
+}
+
+func (mgr *NetworkManager) resolveHostnameGateways(gatewaySet map[NetworkGateway]struct{}) {
+	// filter the list of gateways to resolve
+	hostnameGateways := map[string][]NetworkGateway{}
+	names := sets.NewSet()
+	for gw := range gatewaySet {
+		if gwIP := net.ParseIP(gw.Addr); gwIP != nil {
+			continue
+		}
+		delete(gatewaySet, gw)
+		if !features.ResolveHostnameGateways {
+			log.Warnf("Failed parsing gateway address %s from Service Registry. "+
+				"Set RESOLVE_HOSTNAME_GATEWAYS on istiod to enable resolving hostnames in the control plane.",
+				gw.Addr)
+			continue
+		}
+		hostnameGateways[gw.Addr] = append(hostnameGateways[gw.Addr], gw)
+		names.Insert(gw.Addr)
+	}
+
+	// resolve each hostname
+	for host, addrs := range mgr.NameCache.Resolve(names) {
+		gwsForHost := hostnameGateways[host]
+		if len(addrs) == 0 {
+			log.Warnf("could not resolve hostname %q for %d gateways", host, len(gwsForHost))
+		}
+		// expand each resolved address into a NetworkGateway
+		for _, gw := range gwsForHost {
+			for _, resolved := range addrs {
```

Comment:
A bit confused by this. So if there are multiple addrs, we just use the last one? Especially concerned with this WRT to the A/AAAA queries - its plausible the LB is available under both but the cluster is only v4/v6 so one or the other will fail?

Reply:
Also ideally long term we can have multiple IPs for a single gw and adjust the weight proportionally? But certainly not a P0

Reply:
This expands into a gateway instance per address in the answer. The weighting logic is already handled in our EDS code.

Reply:
I am not sure I understand. For each iteration of `addrs`, we set `resolvedGw.Addr = resolved`. Won't that just overwrite each time and leave only the last `addrs`?